sbol2/componentdefinition.py

```diff
@@ -134,7 +129,46 @@ def __init__(self, uri=URIRef("example"),
         self.sequenceConstraints = OwnedObject(self, SBOL_SEQUENCE_CONSTRAINTS,
                                                SequenceConstraint,
                                                '0', '*', None)
-        self._hidden_properties = [SBOL_SEQUENCE]
+        # This caches a sequence object so that it can be added
+        # to a document later
+        self._sequence_cache: Union[Sequence, None] = None
+
+    @property
+    def sequence(self):
+        seqs = self.sequences
+        if not seqs:
+            return None
+        seq_uri = seqs[0]
+        if self._sequence_cache and self._sequence_cache.identity == seq_uri:
+            return self._sequence_cache
+        # Not in the cache, try to look up in the document
+        if self.doc:
+            try:
+                return self.doc.find(seq_uri)
+            except NotImplementedError:
+                raise
+
+    @sequence.setter
+    def sequence(self, sequence: Union[Sequence, None]):
+        if not sequence:
+            # TODO: How to remove an orphaned sequence from the document?
+            self.sequences = None
+            self._sequence_cache = None
+            return
+        if self.doc:
+            try:
+                self.doc.add(sequence)
+            except SBOLError as e:
+                if e.error_code() != SBOLErrorCode.DUPLICATE_URI_ERROR:
+                    raise
+        self._sequence_cache = sequence
+        self.sequences = [sequence.identity]
+
+    def added_to_document(self, doc):
+        super().added_to_document(doc)
+        # Add the sequence to the document
+        if self._sequence_cache:
+            doc.add(self._sequence_cache)
```

Comment:
Do we need a check for removal as well?

Reply:
I was wondering the same thing myself. This is how I finessed adding the `Sequence` to the `Document` when the `ComponentDefinition` is added to the `Document`. We have a unit test for that.

I'm not sure that makes sense in a pure SBOL world, but I'd bet there are programs out there that rely on that capability.

Reply:
I'm reluctant to add a `SBOLObject.removed_from_document` method because we have no use for it right now.

Reply:
The edge case I'm thinking about is `test_remove_hidden_sequence` in `test_componentdefinition.py`.  Does this test pass given your changes above?  If not, then there's probably some extra handling that can go here to deal with it.

Reply:
Use a friend class!  (J/k, that's just a little C++ humor to brighten your day)

Reply:
`test_remove_hidden_sequence` in `test_componentdefinition.py` does pass with these changes. I removed the `@unittest.expectedFailure` as part of these changes.

I think what makes this work is the lookup via the document, which ignores a cached sequence. So once the Sequence is removed from the document, the ComponentDefinition can no longer look it up.

Reply:
Cool.  That's elegant.